tool/populate_pub_cache.dart

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+final _client = HttpClient();
+
+/// Fetches and populates a local .pub-cache directory without using `pub`
+/// client and `pub.dev` services.
+Future<void> main(List<String> args) async {
+  final baseDir = File(Platform.script.toFilePath()).parent.parent;
+  print('Base directory: $baseDir');
+
+  final pubCacheDir = Directory('${baseDir.path}/build/pub-cache');
+  print('PUB_CACHE_DIR: ${pubCacheDir.path}');
+  pubCacheDir.createSync(recursive: true);
+
+  final multiVersions = <String, Set<String>>{};
+  final lockFiles = baseDir
+      .listSync(recursive: true)
+      .whereType<File>()
+      .where((f) => !f.path.contains('/pkg/code_coverage/'))
+      .where((f) => !f.path.contains('/pkg/pub_integration/'))
+      .where((f) => f.path.endsWith('pubspec.lock'));
```

Comment:
hmm, what if we passed the list of lockfiles as arguments?

Or if we rename: `tool/populate_pub_cache.dart` to `tool/pub_get_offline.dart` and make it wrap `dart pub get --offline --no-precompile`.
So that we don't call ` dart pub get --offline --no-precompile` from `Dockerfile`, but only call `tool/pub_get_offline.dart` (for each package we want to `pub get`).

We can choose not to fetch a package if the `<package>-<version>/` folder already exists.

-------------------------

I'm mostly concerned that right now the list of packages we want to fetch dependencies for is specifies in two places: 
 * `tool/populate_pub_cache.dart`, and,
 * `Dockerfile`